k2/csrc/fsa.h

```diff
@@ -18,11 +18,12 @@ using StateId = int32_t;
 using Weight = float;
 
 enum {
-  kFinalSymbol = -1,  // final-costs are represented as arcs with
-                      // kFinalSymbol as their label, to the final
-                      // state (see Fst::final_state).
-  kEpsilon = 0        // Epsilon, which means "no symbol", is numbered zero,
-                      // like in OpenFst.
+  kFinalSymbol = -1,     // final-costs are represented as arcs with
+                         // kFinalSymbol as their label, to the final
+                         // state (see Fst::final_state).
+  kEpsilon = 0,          // Epsilon, which means "no symbol", is numbered zero,
+                         // like in OpenFst.
+  kInvalidStateId = -1,  // it MUST be less than zero, like in OpenFst.
```

Comment:
maybe it's a bad idea to define two enums with the same value, this enables users to call `some_symbol == kInvalidStateId` successfully even if they should check `some_symbol == kFinalSymbol`

Reply:
I expect some compilers won't let you declare conflicting enum's.  In any case, my own preference would just be to use a literal -1 for "invalid state id", and to test it with "if (states[i] < 0) { ... }" which should compile to less code than "if (states[i] == -1)", I'd expect.

Reply:
actually I don't think so, as if we define `-2, -3..` in this enum in the future and users misue the value to initlialize `states[i] = -2` , check `states[i] < 0` will still pass..

BTW, I don't think `==-1` and `<0` will make a big difference in time consumed (I guess for most compilers, they all just call one line CPU instruction)

Reply:
To my best knowledge, you **can** define multiple enum entries with the same value.
It is a bug if a compiler does not support it.

----

In kaldi, there is

```cpp
static const int32 kNoPdf = -1;
```
https://github.com/kaldi-asr/kaldi/blob/master/src/hmm/hmm-topology.h#L86

----

A named constant improves readability.

Reply:
At that time I was coding more like OpenFST.  But my decision on this is: no named constant.  Just use -1.

Reply:
Fine, I will change it.

Reply:
Removed and fixed.